### ISSUE TITLE
Hisense U6H - magiskinit fails b/c bootloader adds skip_initramfs on 2SI device

### ISSUE THREAD

**Issue by latelylk** [topjohnwu/Magisk#8063]
Device: Hisense U6H
Android version: 11
Magisk version name: 81658d4
Magisk version code: 27002

In October 2023 I was successful in patching the boot image, but faced a black screen and no OS when trying to boot from it. The debug logs showed that Magisk was seeing `skip_initramfs` in the patched image of our 2SI device:

```cpp
[ 5.742326] magiskinit: Device config:
[ 5.746102] magiskinit: skip_initramfs=[1]
[ 5.750236] magiskinit: force_normal_boot=[0]
```

Magisk normally patches `skip_initramfs` out of the boot image. At some point in the past devices (ex: Galaxy A21s) started having `skip_initramfs` added to the dtb bootargs. Magisk v24.0 fixed this by [patching the dtb](https://github.com/topjohnwu/Magisk/blob/v24.0/native/jni/magiskboot/dtb.cpp#L143-L149). Similar to how that issue presented, magiskinit treats the Hisense U6H as a `Legacy SAR` device because `skip_initramfs` is added by U-Boot (MBOOT).

Special thanks to @throwaway96 for finding this. Hisense adds `skip_initramfs` in the bootloader (0x274F108C):
```cpp
void env_add_boot_skip_initramfs(void)
{
    const char boot_skip_initramfs[] = "skip_initramfs";
    env_add_bootargs(boot_skip_initramfs, boot_skip_initramfs);
}
```

it does this with a check in the do_readKL() function (0x274F1654):
```cpp
// append command line.
if(strcmp(partition_name,"boot") == 0)
{
    env_add_boot_skip_initramfs();
}
```

Luckily with this understanding we can easily patch Magisk to achieve root:
<details>
  <summary>Patching Magisk :)</summary>
Prereqs:
- [Magisk v27.0 Source](https://github.com/topjohnwu/Magisk/releases/tag/v27.0)*
- Ability to compile Magisk

1. Edit [lines 85-87 of `native/src/init/init.cpp`](https://github.com/topjohnwu/Magisk/blob/v27.0/native/src/init/init.cpp#L85-L87) from:
```cpp
if (config.skip_initramfs)
  init = new LegacySARInit(argv, &config);
else if (config.force_normal_boot)
```
to:
```cpp
/*if (config.skip_initramfs)
    init = new LegacySARInit(argv, &config);
else*/ if (config.force_normal_boot)
```

2. That's it. Compile and use Magisk as normal. Have fun?
</details>

A similar issue to this was created when Galaxy A21s devices began including `skip_initramfs` in the dtb ([xda](https://xdaforums.com/t/root-modified-magisk-for-galaxy-a21s-r.4271537/), #4307). In that issue osm0sis stated:
> [Creating a device/board whitelist is not a proper solution](https://github.com/topjohnwu/Magisk/pull/4307#issuecomment-852523866)

I'm not sure what Magisk can do to solve this without hardcoding for these devices. Patching the bootloader seems risky. Depending on how other devices are structured, can the order for checking `skip_initramfs` and `force_normal_boot` be swapped? Any other ideas?

Thank you :)

Quick reference:
[Hisense U6H XDA Thread](https://xdaforums.com/t/root-hisense-u6h-via-modified-magisk.4635687/)
[Patching Magisk for Galaxy A21s XDA Thread](https://xdaforums.com/t/root-modified-magisk-for-galaxy-a21s-r.4271537/)
#4307

tl;dr Assumptions that Magisk makes should be correct, but the patches Hisense have applied to U-Boot make that not so

**Comment by yujincheng08** [comment]
How about moving

https://github.com/topjohnwu/Magisk/blob/c6f076251033c46a66b8f06a517cd58d94ad2fb7/native/src/init/init.cpp#L96-L97

before

https://github.com/topjohnwu/Magisk/blob/c6f076251033c46a66b8f06a517cd58d94ad2fb7/native/src/init/init.cpp#L90-L91

But I dont know if this will break other devices.

**Comment by latelylk** [comment]
Hi @yujincheng08 just checking whether this is truly "not planned" or if it was closed for being stale. I'll have a device on-hand again soon - am I able to re-open this or open a new issue when I am available to work on it?

**Comment by yujincheng08** [comment]
it was closed for being stale

**Comment by latelylk** [comment]
Perfect. Thank you :)

**Comment by vvb2060** [comment]
try https://install.appcenter.ms/users/vvb2060/apps/magisk/distribution_groups/public
